### ISSUE TITLE
Some (all?) AGC I/O channels should be accessible via multiple addresses

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#1054]
I noticed something interesting today while looking at the schematics for the Service Gates module in the AGC.
![rqg_gate](https://user-images.githubusercontent.com/94056/34400865-878f0d08-eb4a-11e7-95c7-252129003bbe.png)

This gate, 33409, is one of the possible ways for RQG/ (read Q register, gated) signal to be generated. Specifically, this is the gate that enables Q to be read as I/O channel 2. It's effectively acting as an AND gate, whose inputs are:
* RCHG/ -- read an I/O channel, gated with the RT (read time) pulse.
* XB2/ -- the lowest digit of the target address (in octal) is a 2, i.e. the address must be xx**2** for a channel.
* XT0/ -- the second lowest digit of the target address (in octal) is a 0, i.e. the address must be x**0**x for a channel.

Notably missing from this gate, and seemingly a lot of others like it, is any signal indicating what the highest digit must be. In other words, even though the theoretical range for valid channel addresses is 000 through 777, the hardware isn't actually looking at the first digit of the address. So in this case, Q is accessible as channels 102, 202, 302, 402, 502, 602, and 702 in addition to being channel 2 as intended.

I threw together a quick program to test this, and ran it in the hardware sim:
![rqg](https://user-images.githubusercontent.com/94056/34401026-2b763c2e-eb4c-11e7-9920-210084726939.png)

It simply READs channels 2, 102, and 202, clearing A between each. In all three cases, RQG/ is generated and the contents of Q are read into A.

It's worth noting that as far as I can tell, there isn't any particular reason for this, other than saving chips and pins. The Channel 77 Restart Monitor uses a different mechanism for latching the channel address, since XBx, XTx, and YBx don't go out to the test connector; it instead samples the main write bus when it knows a channel address is being sent around. But even then, it only grabs bits 1-6 as the channel number, so the restart monitor is available as 77, 177, 277, etc. Especially in this case, there's nothing stopping a peripheral designer from using all nine bits and making a channel only accessible as channel, say, 350. But it would only work as intended as long as the lower two digits of the channel number don't collide with an already-existing channel. (You could make a channel 102, if you wanted to, in which case the contents of your new channel would be logically ORed with Q upon read). This is all kind of moot, though; the only peripherals we know about (restart monitor, ACM/ATM) used channels with only two digits.

Long story short, I need to go over all of the channels and make sure this is the case for every channel, and then (probably) tweak yaAGC to disregard bits 7-9 of channel instructions.

**Comment by rburkey2005** [comment]
Great detective work!

However ... this is a case where some care is warranted in fixing up yaAGC.  If you make such a fix, please provide the ability to enable/disable the fix via a command-line option.  In fact, I would prefer for the default to be the current behavior, namely for the yaAGC by default to accept all i/o channels as distinct in spite of the hardware's treatment of them.  I would suggest having a command-line option "--hardware-channels" to enable the fix.

The rationale for this is that while it's important to be able to have the yaAGC mimic the hardware behavior as much as possible to purposes of validation of hardware models, it's not yaAGC's only purpose.  One of yaAGC's other purposes is to allow people to have the "AGC software-development experience".  For that purpose, it's necessary for users to be able to write new AGC software for their own whims, and that entails also being able to define their own peripheral devices.  The latter enevitably entails having those peripheral devices use otherwise-unused portions of i/o-channel address space.  But this ability would be greatly hampered if the number of available channels was cut down a lot, or if the address space was full of holes.

I don't know how often this capability is taken advantage of, but I can say that in the last couple of weeks I myself have written the code for a new peripheral device (https://github.com/virtualagc/virtualagc/blob/master/piPeripheral/piPeripheral.py), along with an AGC program for accessing it (https://github.com/virtualagc/virtualagc/blob/master/piPeripheral/piPeripheral.agc), along with a general description on the website for how to do such stuff.  This new peripheral uses a bunch of unassigned i/o channels, 040-071.  The new channels encompass stuff like GPS, magnetometer, real-time clock, and so on. It's easy to imagine other fanciful peripherals that use a bunch more channels.  The existence of physical DSKYs with embedded yaAGC on a Raspberry Pi platform that has many available hardware plug-in devices opens up the possibility of many such fanciful peripheral devices.

**Comment by thewonderidiot** [comment]
That's certainly fair. How many lost channels would be too many? A proper fix for this only removes approximately a third of the unused channels -- the number would drop from 491 (decimal) available channels to 344 (decimal) unused channels, which is still a decent bit. Admittedly there would be annoying gaps, but we could have big notices on the development pages advising to stay in the x20-x27 and x40-x76 area for selecting new channels. We could even make yaYUL generate noisy warnings if code is assembled that attempts to access affected channels.

**Comment by rburkey2005** [comment]
Well, as we all know, 640K should be enough memory for anybody.  :smile:

But I guess I had misunderstood you to say that the total number of channels would drop from 512 to 64, more or less, but that more analysis would be needed to get the exact numbers.  I would say that 64 total is too few.  If the number is really >300, I don't have any problem with that at all.

**Comment by thewonderidiot** [comment]
Hehehe. Yeah, I worded that poorly, sorry. I meant making it disregard bits 7-9 for affected channels. Including the Restart Monitor, the AGC uses 21 of the 512 possible channels. In the worst possible case of all of these channels being affected (which is looking to likely be the case), each of those would be have eight copies, one for each different combination of bits 7-9. So the total number of unused channels would drop from `512 - 21 = 491` to `512 - 8*21 = 344`.

One possible thing we could do to lessen the impact even more would be to relocate our reserved "fake" channels to duplicated copies of the AGC-internal channels (e.g. use 102 instead of 163), since those would almost certainly go unused by peripherals. That would be a breaking change to anybody using the existing fake channels, though, so it might not be a good idea.

**Comment by rburkey2005** [comment]
Given those numbers, I don't think the impact needs to be lessened.  As far as moving the fictitious channels is concerned, that's basically a published API.  We don't know who's using that API (possibly nobody, but possibly many people), so while it may be technically possible to change it, I don't think it's too palatable a notion in reality.

**Comment by rburkey2005** [comment]
I've created a new branch, Fictitious-1201, with the following experimental code:

- A CLI switch for yaYUL, --bailout=_BailoutName_, which puts a `TC BailoutName` at address 07776.  In other words, it prepares an AGC rope for use with the fictitious input channel 162 I proposed earlier.
- Luminary099 is assembled using --bailout=BAILOUT1.
- Luminary099.binsource is similarly modified (05616 is put at 07776, which is normally @).
- All other AGC versions are left unchanged.
- yaAGC is modified to have a fictitious channel 162, for which: if 020000 is written to the channel, output messages channels 010 and 011 are subsequently blocked, so the DSKY never updates; this block is removed when anything else is written to the channel.  If 01104, 01201, 01202, 01203, or 01211 is written to the channel, this value is deposited in address 07777 in memory, and program control is diverted to address 07776, though this isn't done unless the conditions are also appropriate for an interrupt to occur; i.e., it doesn't happen during an interrupt, EXTEND, or INDEX, or various other conditions.  Thus, a bailout with one of those numbers as the program alarm occurs.
- A Python program, fakeAlarms.py, has been added to the piPeripheral directory.  This program performs a one-shot write to any channel of the AGC.  You can see the full options with "fakeAlarms.py --help", but the default channel it writes to is 162 and the default value it writes is 01201, so if used without arguments it triggers a 1201.  The following sequence would freeze a DSKY and then 10 seconds later send a 1201 bailout command to a running copy of yaAGC:

 fakeAlarms.py --value=20000
 sleep 10
 fakeAlarms.py

Similarly, the following sequence would trigger a 1202 alarm:

 fakeAlarms.py --value=1202

It seems to work, though in the limited fashion I'm capable of using it, it isn't too impressive.  At any rate, it's there to experiment with if anybody's interested.
